src/Lean/Meta/IndPredBelow.lean

```diff
@@ -374,6 +378,183 @@ where
+    let newAlt ← mkLambdaFVars (fvars[:oldFVars.size] ++ xs[oldFVars.size:] ++ fvars[oldFVars.size:]) t
+    trace[Meta.IndPredBelow.match] "alt {idx}:\n{alt} ↦ {newAlt}"
+    newAlt
+
+  let matcherName ← mkFreshUserName mkMatcherInput.matcherName
+  withExistingLocalDecls (lhss.foldl (init := []) fun s v => s ++ v.fvarDecls) do
+    for lhs in lhss do
+      trace[Meta.IndPredBelow.match] "{←lhs.patterns.mapM (·.toMessageData)}"
+  let res ← Match.mkMatcher { matcherName, matchType, numDiscrs := (mkMatcherInput.numDiscrs + 1), lhss }
+  res.addMatcher
+  check res.matcher
+  let args := #[motive] ++ matcherApp.discrs.push below ++ alts
+  let newApp := mkApp res.matcher motive
+  let newApp := mkAppN newApp $ matcherApp.discrs.push below
+  let newApp := mkAppN newApp alts
+  (newApp, res.addMatcher)
+
+where 
+  addBelowPattern (indName : Name) (lhs : AltLHS) : MetaM AltLHS := do
+    withExistingLocalDecls lhs.fvarDecls do
+    let patterns := lhs.patterns.toArray
+    let originalPattern := patterns[idx]
+    let (fVars, belowPattern) ← convertToBelow indName patterns[idx]
+    withExistingLocalDecls fVars.toList do
+    let patterns := patterns.push belowPattern 
+    let patterns := patterns.set! idx (←toInaccessible originalPattern)
+    { lhs with patterns := patterns.toList, fvarDecls := lhs.fvarDecls ++ fVars.toList }
+      
+  /--
+    this function changes the type of the pattern from the original type to the `below` version thereof.
+    Most of the cases don't apply. In order to change the type and the pattern to be type correct, we don't 
+    simply recursively change all occurrences, but rather, we recursively change occurences of the constructor.
+    As such there are only a few cases:
+    - the pattern is a constructor from the original type. Here we need to:
+      - replace the constructor
+      - copy original recursive patterns and convert them to below and reintroduce them in the correct position
+      - turn original recursive patterns inaccessible
+      - introduce free variables as needed.
+    - it is an `as` pattern. Here the contstructor could be hidden inside of it.
+    - it is a variable. Here you we need to introduce a fresh variable of a different type.
+  -/
+  convertToBelow (indName : Name) 
+      (originalPattern : Pattern) : MetaM $ Array LocalDecl × Pattern := do 
+    match originalPattern with
+    | Pattern.ctor ctorName us params fields => 
+      let ctorInfo ← getConstInfoCtor ctorName
+      
+      let belowCtor ← getConstInfoCtor $ ctorName.updatePrefix $ ctorInfo.induct ++ `below
+      let belowIndices ← IndPredBelow.getBelowIndices ctorName
+      let belowIndices := belowIndices[ctorInfo.numParams:].toArray.map (· - belowCtor.numParams)
+
+      let mut belowFieldOpts := mkArray belowCtor.numFields none
+      let fields := fields.toArray
+      for fieldIdx in [:fields.size] do
```

Comment:
Are `belowCtor.numFields` and `fields.size` supposed to be equal?
After the for-loop, is it possible to have `none` in `belowFieldOpts`? The function `transformFields` suggests this is a possibility.

Reply:
see comment.